File: import_ptn_data.js
const https = require('https');

const url = 'https://docs.google.com/spreadsheets/d/1eZHfayqypBQ3XRGf5hRMlIaMKiz0OPylow4T891mqSs/export?format=csv&gid=1813610957';

function fetchCSV(url) {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            let data = '';
            res.on('data', (chunk) => data += chunk);
            res.on('end', () => resolve(data));
            res.on('error', (err) => reject(err));
        });
    });
}

function parseCSV(csv) {
    const lines = csv.split(/\r?\n/);
    const result = {};

    // Header is usually around line 2 (0-indexed). 
    // We'll scan for the header line starting with "PROVINSI" or containing "UNIVERSITAS"
    let headerIndex = -1;
    let uniCol = -1;
    let prodiCol = -1;
    let jenjangCol = -1;

    for (let i = 0; i < Math.min(20, lines.length); i++) {
        const line = lines[i];
        if (line.includes('UNIVERSITAS') && line.includes('PROGRAM STUDI')) {
            headerIndex = i;
            const cols = line.split(','); // Simple split for header check
            // Find indices
            cols.forEach((col, idx) => {
                const c = col.toUpperCase().trim();
                if (c === 'UNIVERSITAS') uniCol = idx;
                if (c === 'PROGRAM STUDI') prodiCol = idx;
                if (c === 'JENJANG') jenjangCol = idx;
            });
            break;
        }
    }

    if (uniCol === -1 || prodiCol === -1) {
        console.error("Could not find headers");
        process.exit(1);
    }

    // Parse specific CSV row logic (handling quotes)
    const parseRow = (row) => {
        const res = [];
        let current = '';
        let inQuote = false;
        for (let i = 0; i < row.length; i++) {
            const char = row[i];
            if (char === '"') {
                inQuote = !inQuote;
            } else if (char === ',' && !inQuote) {
                res.push(current.trim());
                current = '';
            } else {
                current += char;
            }
        }
        res.push(current.trim());
        return res;
    };

    for (let i = headerIndex + 1; i < lines.length; i++) {
        const line = lines[i];
        if (!line || line.trim() === '') continue;

        const cols = parseRow(line);
        if (cols.length <= Math.max(uniCol, prodiCol)) continue;

        const uni = cols[uniCol]?.replace(/^"|"$/g, '').trim();
        const prodi = cols[prodiCol]?.replace(/^"|"$/g, '').trim();
        const jenjang = jenjangCol !== -1 ? cols[jenjangCol]?.replace(/^"|"$/g, '').trim() : '';

        if (!uni || !prodi || uni.toUpperCase() === 'UNIVERSITAS') continue;

        if (!result[uni]) result[uni] = new Set();

        let formatted = prodi;
        // If Jenjang is D3/D4, append it if not already valid
        if (jenjang && ['D3', 'D4', 'D1', 'D2'].includes(jenjang)) {
            formatted = `${prodi} (${jenjang})`;
        } else if (jenjang && jenjang.startsWith('S1') === false && jenjang !== '') {
            // Handle other cases? S1 is default so we don't add it.
            // Maybe "Sarjana Terapan" -> D4
        }

        // Clean up common dirty data
        formatted = formatted.replace(/""/g, '"');

        result[uni].add(formatted);
    }

    return Object.keys(result).map(key => ({
        university: key,
        prodis: Array.from(result[key]).sort()
    })).sort((a, b) => a.university.localeCompare(b.university));
}

fetchCSV(url).then(csv => {
    const data = parseCSV(csv);
    // Be careful with output size. Console.log might truncate.
    // Better to write to a file.
    const fs = require('fs');
    const content = `const masterDataPTN = ${JSON.stringify(data, null, 4)};`;
    fs.writeFileSync('ptn_data.js', content); // Overwrite existing ptn_data.js
    console.log("Successfully updated ptn_data.js with " + data.length + " universities.");
}).catch(err => {
    console.error("Failed:", err);
});

File: import_ptn_data.py
import csv
import json
import urllib.request
import io

URL = "https://docs.google.com/spreadsheets/d/1eZHfayqypBQ3XRGf5hRMlIaMKiz0OPylow4T891mqSs/export?format=csv&gid=1813610957"
OUTPUT_FILE = "ptn_data.js"

def fetch_csv(url):
    print(f"Fetching CSV from {url}...")
    with urllib.request.urlopen(url) as response:
        return response.read().decode('utf-8')

def parse_csv(csv_content):
    print("Parsing CSV...")
    f = io.StringIO(csv_content)
    reader = csv.reader(f)
    
    headers = []
    data = {}
    
    # Locate header row
    uni_idx = -1
    prodi_idx = -1
    jenjang_idx = -1
    
    row_count = 0
    start_parsing = False
    
    for row in reader:
        row_count += 1
        if not row: continue
        
        # Header detection logic
        if not start_parsing:
            # Check if this row looks like a header
            row_upper = [c.upper().strip() for c in row]
            if "UNIVERSITAS" in row_upper and "PROGRAM STUDI" in row_upper:
                print(f"Found header at row {row_count}")
                try:
                    uni_idx = row_upper.index("UNIVERSITAS")
                    prodi_idx = row_upper.index("PROGRAM STUDI")
                    if "JENJANG" in row_upper:
                        jenjang_idx = row_upper.index("JENJANG")
                    start_parsing = True
                    continue
                except ValueError:
                    continue
        
        if start_parsing:
            if len(row) <= max(uni_idx, prodi_idx): continue
            
            uni = row[uni_idx].strip()
            prodi = row[prodi_idx].strip()
            jenjang = row[jenjang_idx].strip() if jenjang_idx != -1 and len(row) > jenjang_idx else ""
            
            if not uni or uni.upper() == "UNIVERSITAS": continue
            if not prodi: continue
            
            # Clean Uni Name
            uni = uni.replace('"', '').strip()
            
            # Key Formatting
            if uni not in data:
                data[uni] = set()
            
            # Prodi Formatting
            formatted_prodi = prodi
            if jenjang and jenjang.upper() in ["D3", "D4", "D1", "D2"]:
                formatted_prodi = f"{prodi} ({jenjang})"
            
            data[uni].add(formatted_prodi)
            
    return data

def main():
    try:
        csv_content = fetch_csv(URL)
        parsed_data = parse_csv(csv_content)
        
        # Convert to final structure
        result_list = []
        for uni in sorted(parsed_data.keys()):
            prodis = sorted(list(parsed_data[uni]))
            result_list.append({
                "university": uni,
                "prodis": prodis
            })
            
        print(f"Processed {len(result_list)} universities.")
        
        # Write to JS file
        js_content = f"const masterDataPTN = {json.dumps(result_list, indent=4)};"
        
        with open(OUTPUT_FILE, "w", encoding="utf-8") as f:
            f.write(js_content)
            
        print(f"Successfully wrote data to {OUTPUT_FILE}")
        
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()
